app/src/main/java/de/rochefort/childmonitor/ListenService.kt

```diff
@@ -0,0 +1,212 @@
+        // Tell the user we stopped.
+        Toast.makeText(this, R.string.stopped, Toast.LENGTH_SHORT).show()
+    }
+
+    override fun onBind(intent: Intent): IBinder {
+        return binder
+    }
+
+    private fun buildNotification(name: String?): Notification {
+        // In this sample, we'll use the same text for the ticker and the expanded notification
+        val text = getText(R.string.listening)
+
+        // The PendingIntent to launch our activity if the user selects this notification
+        val contentIntent = PendingIntent.getActivity(this, 0,
+                Intent(this, ListenActivity::class.java), PendingIntent.FLAG_IMMUTABLE)
+
+        // Set the info for the views that show in the notification panel.
+        val b = NotificationCompat.Builder(this, CHANNEL_ID)
+        b.setSmallIcon(R.drawable.listening_notification) // the status icon
+                .setOngoing(true)
+                .setTicker(text) // the status text
+                .setContentTitle(text) // the label of the entry
+                .setContentText(name) // the contents of the entry
+                .setContentIntent(contentIntent)
+        return b.build()
+    }
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val serviceChannel = NotificationChannel(
+                    CHANNEL_ID,
+                    "Foreground Service Channel",
+                    NotificationManager.IMPORTANCE_DEFAULT
+            )
+            notificationManager!!.createNotificationChannel(serviceChannel)
+        }
+    }
+
+    fun setErrorCallback(errorCallback: Runnable?) {
+        mErrorCallback = errorCallback
+    }
+
+    fun setUpdateCallback(updateCallback: Runnable?) {
+        mUpdateCallback = updateCallback
+    }
+
+    inner class ListenBinder : Binder() {
+        val service: ListenService
+            get() = this@ListenService
+    }
+
+    private var mErrorCallback: Runnable? = null
+    private var mUpdateCallback: Runnable? = null
+    private fun doListen(address: String?, port: Int) {
+        val lt = Thread {
+            try {
+                val socket = Socket(address, port)
+                streamAudio(socket)
+            } catch (e: IOException) {
+                Log.e(TAG, "Failed to stream audio", e)
+            }
+            if (!Thread.currentThread().isInterrupted) {
+                // If this thread has not been interrupted, likely something
+                // bad happened with the connection to the child device. Play
+                // an alert to notify the user that the connection has been
+                // interrupted.
+                playAlert()
+                val errorCallback = mErrorCallback
+                errorCallback?.run()
+            }
+        }
+        this.listenThread = lt
+        lt.start()
+    }
+
+    @Throws(IllegalArgumentException::class, IllegalStateException::class, IOException::class)
```

Comment:
It's unidiomatic to throw Exceptions in Kotlin.

See e.g. https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07

Reply:
I've wrapped the library functions that throw exceptions in `runCatching` blocks.

Reply:
I personally find the runCatching stuff in combination with result.isFailure more clumsy than try - catch.
Therefore  I don't mind having try catch blocks when calling external api that throws. I am just saying that we should just not propagate exceptions that are part of normal runtime situations. Instead, we should return a value from which we the caller can derive error states.
The behavior is changed now in that streamAudio only returns instead of propagating error state to the caller. This is mostly okay in this particular case because the caller does not need a return value, but we are still losing the "connection failed" logging in the current code state

Reply:
> I personally find the runCatching stuff in combination with result.isFailure more clumsy than try - catch.

I agree, I misunderstood you then.

> This is mostly okay in this particular case because the caller does not need a return value, but we are still losing the "connection failed" logging in the current code state

Sure, but what are you going to do with that information?
On top of that, is it an error, when I close the monitor on the child device by stopping the child activity?
It closes the connection though.

I've re-introduced the log  message though.

Reply:
The point I was trying to make is that your refactoring has changed the error handling logic.
It's okay to do that, but it should be in a separate step and not as part of a refactoring.
This is useful when tracking down bugs using git bisect.
Let me give you an example of changed error handling:
Previously, if the remote address was unreachable, the instantiation of the  would have thrown a SocketException, which in turn would have been caught by a try catch within the Thread block:
```
        val lt = Thread {
            try {
                val socket = Socket(address, port)
                streamAudio(socket)
            } catch (e: IOException) {
                Log.e(TAG, "Failed to stream audio", e)   // <-- here!
            }
```
Now that the handling of that error is no longer defined.
The change I was looking for when I made the original comment was a refactoring that returns Results where Exceptions were thrown previously, and handles failed results where exceptions would previously have been caught.
If it is still unclear, feel free to just revert the changes to the exception handling and I will fix it up to my taste after merging.